src/relations/db_router.py

```diff
@@ -0,0 +1,254 @@
+                    f"Missing information to creata a database and scoped user for {application_name}"
+                )
+                continue
+
+            requested_users.append(RequestedUser(application_name, username, hostname, database))
+
+        return requested_users
+
+    def _create_requested_users(
+        self, requested_users: List[RequestedUser], user_unit_name: str
+    ) -> Tuple[Dict[str, str], Set[str]]:
+        """Create the requested users and said user scoped databases.
+
+        Args:
+            requested_users: A list of RequestedUser
+                (named tuples containing user and database info)
+            user_unit_name: Name of unit from which the requested users will be accessed from
+
+        Returns:
+            tuple containing a dictionary of application_name to password
+                and a list of requested user applications
+
+        Raises:
+            MySQLCheckUserExistenceError if there is an issue checking a user's existence
+            MySQLConfigureRouterUserError if there is an issue configuring the mysqlrouter user
+            MySQLCreateApplicationDatabaseAndScopedUserError if there is an issue creating a
+                user or said user scoped database
+        """
+        user_passwords = {}
+        requested_user_applications = set()
+
+        for requested_user in requested_users:
+            password = self._get_or_set_password_in_peer_databag(requested_user.username)
+
+            if not self.charm._mysql.does_mysql_user_exist(
+                requested_user.username, requested_user.hostname
+            ):
+                if requested_user.application_name == "mysqlrouter":
+                    self.charm._mysql.configure_mysqlrouter_user(
+                        requested_user.username, password, requested_user.hostname, user_unit_name
+                    )
+                else:
+                    self.charm._mysql.create_application_database_and_scoped_user(
+                        requested_user.database,
+                        requested_user.username,
+                        password,
+                        requested_user.hostname,
+                        user_unit_name,
+                    )
+
+            user_passwords[requested_user.application_name] = password
+            requested_user_applications.add(requested_user.application_name)
+
+        return user_passwords, requested_user_applications
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Copy data from the relation's application databag to the leader unit databag
+        since legacy applications expect credential data to be populated on the leader
+        unit databag.
+        """
+        relation = self.charm.model.get_relation(LEGACY_DB_ROUTER)
+        if not relation:
+            return
+
+        relation_databag = relation.data
+        for key, value in relation_databag[self.charm.app].items():
```

Comment:
Just got another case: when syncing, it's possible that the departed unit was db primary (`db_host`), so this needs to be updated also.

Reply:
Good catch - I was setting the `db_host` as the leader unit's address rather than the cluster primary. Addressed in [b4f569d](https://github.com/canonical/mysql-operator/pull/15/commits/b4f569df9c33bec957c9f3eafd520c9950054b00)